README.md
# apt-spy


debian/changelog
apt-spy (3.2.2-2) unstable; urgency=low

  * built by OLS3 for used locally.

apt-spy (3.2.2-1) unstable; urgency=low

  * New upstream release
   - Fixes segfault with option -m and special args (Closes: #645839)
   - Fixes another segfault (Closes: #447232)
   - Fixes segfault with with -v or -h options (Closes: #491802)
   - Fixes the region filter (Closes: #617699)
   - Fixes paths in the man page (Closes: #548591, #551344)
   - Doesn't get confused by leading blanks (Closes: #457049)
   - Inlines initialization of defaults (Closes: #317592)
   - Tests all of the IP addresses for a mirror (Closes: #320112)
  * Applied patch to copyright and other files (Closes: #645910)
  * Moved mirrors.txt from /var/cache/apt-spy to /var/cache/apt
  * Updated to latest policy, ???no changes needed?????
  * Switch to dpkg-source 3.0 (quilt) format

 -- Stefano Canepa <sc@linux.it>  Sat, 14 Jan 2012 00:02:21 +0100

apt-spy (3.1-19) unstable; urgency=low
  
  * Package adopted (Closes: #405297)
  * Merged all patches from debian/patches. Quilt build-dependency dropped.
  * Moved mirror.txt from /var/lib/apt-spy/ to /var/cache/apt-spy 
    (Closes: #317587)
  * Included command line used to generate source.list (Closes: #432592)
  * Added -y option to choose sections other than main (Closes: #464555)
  * Upload sponsored by Emanuele Rocca <ema@debian.org>
  
 -- Stefano Canepa <sc@linux.it>  Sat, 12 Jul 2008 16:03:37 +0200

apt-spy (3.1-18) unstable; urgency=low

  * QA upload.
  * Depend on libcurl4-openssl-dev instead of libcurl3-openssl-dev
     Closes: #452000
  * Bumped Standards-Version: 3.7.3, no change needed
  * Removed usr/share/lintian/overrides/ from debian/dirs (unused)

 -- Raphael Geissert <atomo64@gmail.com>  Tue, 18 Dec 2007 17:55:18 -0600

apt-spy (3.1-17) unstable; urgency=low

  * QA upload.
  * Updated default mirrors list (new 02_mirrors-txt_update.diff)
  * Removed debian/TODO.Debian since there's nothing left to do
  * argument-parsing-error-checks.diff: improved error checking on passed arguments (Closes: #317590)
    - Thanks to Justin Pryzby
  * segfault_418548.diff: invalid pointer causes a segfault (Closes: #418548)
    - Thanks to Steve Kemp for the patch
  * sources.list-rename.diff: default output list to /etc/apt/sources.list.d/apt-spy.list avoiding overwrite of /etc/apt/sources.list (Closes: #442165)
  * Updated debian/menu to reflect new Debian Menu Sub-Policy

 -- Raphael Geissert <atomo64@gmail.com>  Sat, 20 Oct 2007 21:05:46 -0500

apt-spy (3.1-16) unstable; urgency=low

  * QA upload.
  * 03_writetop_segfault.diff: Fix annoying segfault when the -w option
    is used (closes: #394376).
  * 04_http_missing.diff: Make parser not to fail when HTTP field 
    is missing (closes: #389838).
  * 02_mirrors-txt_update.diff: Update README.mirrors.txt.

 -- Robert Luberda <robert@debian.org>  Tue, 27 Feb 2007 23:30:46 +0100

apt-spy (3.1-15) unstable; urgency=low

  * QA upload.
  * Orphaning package: thanks Stephen Stafford for your work.
     - change debian/control accordingly.
     - ACK previous NMU.
  * Fix typo in apt-spy.8 manual page: (Closes: #401985)
     - new 01_typo_fix_apt-spy8.diff patch.
     - Build-Depend on quilt and include quilt's makefile in rules.
  * Use dh_installman and dh_install instead of copying.
  * Get rid of lintian override.
  * Change cosmetics in rules' update target.
  * Update mirrors.txt file as 02_mirrors-txt_update.diff.
  * Add a TODO.Debian file.

 -- Mohammed Adnène Trojette <adn+deb@diwi.org>  Tue,  2 Jan 2007 19:21:19 +0100

apt-spy (3.1-14.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Mirrors now use ls-LR.gz. Modify main.c accordingly. (Closes: #356654)
  * Bump Standards-Version to 3.7.2 and et debhelper compatibility to 5:
     + remove debian/conffiles.
     + add binary-indep target to rules.
     + add debian/compat.

 -- Mohammed Adnène Trojette <adn+deb@diwi.org>  Thu, 28 Sep 2006 03:44:36 +0200

apt-spy (3.1-14) unstable; urgency=low

  [ ssta ]
  * Change build-depends on libcurl3-dev to libcurl3-openssl-dev
  * Apply some patches from Justin Pryzby (Thanks Justin, apologies it's
    taken me so long!)
    - Patch to fix the mirror file screwup which caused apy-spy to segfault
      every time. It removes some arbitrary limits. (Closes: #315120 #317591)
    - Change the malloc of country_file so it's not mallocing its own length
      (Closes: #317585)
    - Apply patch which uses sprintf instead of strncat in file.c
      (Closes: #317589)
    - Apply patch to fix small memory leak (Closes: #317593)
    - s/LICENSE/COPYRIGHT/ in *.c  (Closes: #317588)
    - add a break statement to write_top() (Closes: #317586)

  [ Stephen Stafford ]
  * 

 -- Stephen Stafford <bagpuss@debian.org>  Sun, 16 Oct 2005 12:42:22 +0000

apt-spy (3.1-13) unstable; urgency=low

  * Transition to libcurl3 (Closes: #279472)
  * fix manpage (Closes: #281897)

 -- Stephen Stafford <bagpuss@debian.org>  Thu,  9 Dec 2004 12:45:08 +0000

apt-spy (3.1-12) unstable; urgency=low

  * file.c: remove assumption of signedness of char when comparing to EOF.
    This had the effect of causing the thing to hang on powerpc.  Remove the
    cast altogether as getc() returns an int anyway.  (Closes: #282121)
  * Update README.mirrors.txt to October 26th version.

 -- Stephen Stafford <bagpuss@debian.org>  Mon, 22 Nov 2004 10:39:00 +0000

apt-spy (3.1-11) unstable; urgency=low

  * move README.mirrors.txt to /var/lib/apt-spy.  I still am unconvinced,
    but I appear to be in the minority.  (Closes: #277816)
  * rename README.mirrors.txt to mirrors.txt to make it look less like a
    documentation file.  (Closes: #278201)
  * update README.mirrors.txt to Oct13 version.

 -- Stephen Stafford <bagpuss@debian.org>  Tue, 26 Oct 2004 16:56:49 +0100

apt-spy (3.1-10) unstable; urgency=medium

  * Backing out the previous NMU of Martin's.
    apt-spy does not "dynamically update".  It updates *if and ONLY if*

main.c
/*
 * apt-spy (c) Steven Holmes, 2003.
 *
 * This software is licensed as detailed in the COPYRIGHT file
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <limits.h>
#include  <sys/stat.h>

#include "include/main.h"
#include "include/update.h"
#include "include/file.h"
#include "include/parse.h"
#include "include/benchmark.h"
#include "include/global.h"

int bestnumber=BESTNUMBER;

int main(int argc, char *argv[])
{
	int c; /* used to parse arguments */
	char *cur_entry; /* Entry we are benchmarking */

	char *distrib = NULL; /* distribution to use. */
	char *mirror_list = D_MIRROR; /* mirror list file */
	char *config_file = D_CONFIG; /* configuration file */
	char *proxy = NULL; /* Proxy server to use */
	char *infile = NULL; /* optional infile */
	char *outfile = D_OUT; /* outfile name */
	char *topfile = NULL;
	char *area = strdup(D_AREA); /* Area to test */
	char *grab_file = D_FILE; /* File to grab */
	char *update_url = D_UPDATE_URL; /* URL to use for updating */
	char *country_list = NULL; /* List of countries to b/m */
	char *section_list = NULL; /* List of section to include */
	FILE *infile_p, *outfile_p;	/* input/output file pointers */
	FILE *config_p; /* config file pointer */
	FILE *mirror_p; /* mirror list pointer */
	FILE *topfile_p;
	int timeout = 15; /* time to benchmark each server */
	/* int toplist = 0; *//* Whether to write toplist. *//* UNUSED */
	int argsCount = 0; /* persistent counter for args */
	char *args[100]; /* persistent args array */
	/* char str[100] = ""; *//* We don't need to put optarg in str */
	char str[4] = "-? "; /* short option string */
	/* int i; *//* UNUSED */
	char *end;
	struct stat buf;

	/* Number of servers to test. If negative, test them all. */
	int test_number = -1;

	/* Server information structures. */
	server_t current;
	server_t *best = NULL;

	/* Parse options... */
	while((c = getopt(argc, argv, "y:a:c:d:e:f:i:m:o:p:s:t:u:w:n:vh")) != -1)
	{
		if (optarg)
		{ /* not set with -h */
			str[1] = c;
			/* args[argsCount] = (char*)malloc(strlen(str)); *//* BAD: it should be malloc(strlen(str)+1) */
			/* strcat(str, optarg); *//* BAD: optarg may be too long */
			args[argsCount] = malloc(strlen(str)+strlen(optarg)+1);
			memset(args[argsCount],0,strlen(str)+strlen(optarg)+1);
			strncpy(args[argsCount],str,strlen(str));
			strncat(args[argsCount],optarg,strlen(optarg));
			argsCount++;
		}

		switch(c) {
		 /* Sections to include into apt-source */
		case 'y':
			section_list = optarg;
			break;
		 /* Area to benchmark */
		case 'a':
			free(area); /* allocated by strdup */
			area = optarg;
			break;
		/* Distribution we'll write into apt-sources. */
		case 'd':
			distrib = optarg;
			break;
		/* Configuration file to use */
		case 'c':
			/* printf("Option: %c\n",c);
			printf("Option arg: %s\n", optarg); */
			/* config_file = (char *)malloc(25);
			strcpy(config_file, optarg);
			*/
			config_file = optarg;

			printf("Using configuration file: %s\n", config_file);
			break;
		/* Number of servers to benchmark */
		case 'e':
			test_number = strtol(optarg, &end, 10);
			if (!*optarg || end!=optarg+strlen(optarg)) {
				fprintf(stderr, "Error parsing number"
						" of servers to be"
						" benchmarked\n");
				exit(1);
			}
			break;
		/* File, relative to Debian base, to grab from server. */
		case 'f':
			grab_file = optarg;
			break;
		/* User-specified list of servers to benchmark. */
		case 'i':
			infile = optarg;
			break;
		/* The list of mirrors */
		case 'm':
			mirror_list = optarg;
			break;
		/* The output file we use */
		case 'o':
			outfile = optarg;
			break;
		/* Proxy server we should use */
		case 'p':
			proxy = optarg;
			break;
		/* List of countries to benchmark */
		case 's':
			country_list = optarg;
			break;
		/* Time for which to benchmark each server. */
		case 't':
			timeout = strtol(optarg, &end, 10);
			if (!*optarg || end!=optarg+strlen(optarg)) {
				fprintf(stderr, "Error parsing server"
						" benchmark time"
						" interval\n");
				exit(1);
			}
			break;
		/* The URL we should update ourselves from */
		case 'u':
			update_url = optarg;
			break;
		/* Should we write a list of the "top" servers? */
		case 'w':
			/* toplist = 1; *//* UNUSED */
			topfile = optarg;
			break;
		/* Number of servers to write in "top" server list */
		case 'n':
			bestnumber = strtol(optarg, &end, 10);
			if (!*optarg || end!=optarg+strlen(optarg)) {
				fprintf(stderr, "Error parsing number"
						" of best servers to"
						" write\n");
				exit(1);
			}
			break;
		case 'v':
			version();
			break;
		/* Help!! */
		case 'h':
		default:
			usage();			/* display help */
			break;
		}
	}
	argc -= optind;
	argv += optind;

	/* Simple check for stupidity */
	if ((test_number >= 0) && (bestnumber > test_number))
		bestnumber = test_number;

	best = malloc(sizeof(server_t) * (bestnumber + 1));

	if (best == NULL) {
		perror("malloc");
		exit(1);
	}

	/* Zero the "best" structure (just after malloc for readability) */
	memset(best, 0, sizeof(server_t) * (bestnumber + 1));

	/* We require an area and distribution argument if we are not updating */
	if ((argc == 0) && (distrib == NULL))
		usage();

	/* Check for silly combination of country and area arguments */
/*	if ((area != NULL) && (country_list != NULL))
		usage();
*/
	/* Setup default area argument */
/*	if ((area == NULL) && (country_list == NULL))
		area = d_area;
*/
	/* Setup default file argument if none given */
/*	if (grab_file == NULL)
		grab_file = d_file;
*/

	if (strcmp(area, D_AREA) && (country_list != NULL))
		usage();

	/*
	 * if the user does not specify the mirror list file verify if
	 * there is the file and if not download it from main debian
	 * mirror
	 */
/*	if (strcmp(mirror_list, D_MIRROR) == 0) { */
		if (stat(mirror_list, &buf)!= 0) {
			mirror_p = select_mirror(mirror_list, -1);
      if(mirror_p != NULL) {
        if (update(mirror_p, update_url, proxy) != 0) {
          fprintf(stderr, "Failed to download mirror list. Exiting.\n");
          exit(1);
        }
      }
      else
      {
        perror("fopen");
        fprintf(stderr, "Cannot open/create file: %s\n", mirror_list);
        exit(1);
      }
    }
/* } */


	/* Open mirror file. We pass argc so it can tell if we're updating
	   or not */
	mirror_p = select_mirror(mirror_list, argc);
	if (mirror_p == NULL) {
		perror("fopen");
		fprintf(stderr, "Error opening mirror file.\nPerhaps your are missing the \"update\" parameter. Exiting.\n");
		exit(1);
	}

	/* the only possible argument now is "update", for updating the
	   mirrors list */
	if (argc == 1) {
		if (strcmp(argv[0], "update") != 0)
			usage();
/* If necessary, set update_url to default */
		/*	if (update_url == NULL)
			update_url = d_update_url;
		*/
		if (update(mirror_p, update_url, proxy) != 0) {
			fprintf(stderr, "Update failed. Exiting.\n");
			exit(1);
		}
		printf("Update complete. Exiting.\n");
		exit(0);
	}

	/* argc should be 0. If not, there's something wrong. */
	if (argc != 0)
		usage();


	/* We open the infile. Either a temporary file, or a user-specified
	   one. */
	infile_p = select_infile(infile);
	if (infile_p == NULL) {
		perror("tmpfile");
		fprintf(stderr, "Error opening infile (%s). Exiting.\n", infile);
		exit(1);
	}

	/* Set up default if user hasn't specified an outfile */
/*	if (outfile == NULL)
		outfile = d_out;
*/
	/* Open the output file... */
	outfile_p = select_outfile(outfile);
	if (outfile_p == NULL) {
		perror("fopen");
		fprintf(stderr, "Error opening output file (%s). Exiting.\n", outfile);
		exit(1);
	}

	/* Open the topfile */
	if (topfile) {
		topfile_p = select_outfile(topfile);
		if (topfile_p == NULL) {
			perror("fopen");
			fprintf(stderr, "Error opening topfile (%s). Exiting.\n", topfile);
			exit(1);
		}
	}

	/* Open config file */
	config_p = select_config(config_file);
	if (config_p == NULL) {
		perror("fopen");
		fprintf(stderr, "Error opening config file (%s). Exiting.\n", config_file);
		exit(1);
	}

	/* Fill temporary file with useful stuff if it's not user-specified. */
	if (infile == NULL) {
/*		if (area != NULL) {
			if (build_area_file(config_p, infile_p, mirror_p, area) != 0) {
				fprintf(stderr,
				"Error building area file. Exiting.\n");
				exit(1);
			}
		} else {
*/
		if (country_list) {
			if (build_country_file(config_p, infile_p, mirror_p,
			    country_list) != 0) {
				fprintf(stderr,
				"Error building country file. Exiting.\n");
				exit(1);
			}

		} else {
			if (build_area_file(config_p, infile_p, mirror_p, area) != 0) {
				fprintf(stderr,
						"Error building area file. Exiting.\n");
				exit(1);
			}
		}
	}

	/* Make sure we're at the beginning... */
	rewind(infile_p);

	/* This is the main loop. It'll exit when we've exhausted the URL
	   list or test_number is 0 */

	while (((cur_entry = next_entry(infile_p)) != NULL) && test_number != 0) {
		if (ferror(infile_p) != 0) {
			fprintf(stderr, "Error while reading input file\n");
			exit(1);
		}

		/* Turn entry into a pretty structure */
		tokenise(&current, cur_entry);

		/* Do the benchmark */
		if (benchmark(&current, proxy, timeout, grab_file) != 0) {
			fprintf(stderr,
			"Error while performing benchmark. Exiting.\n");
			exit(1);
		}

		decide_best(&current, best);
		free(cur_entry);

		if (test_number > 0)
			--test_number;
	}

	/* write the results */
        printf("Writing new sources.list file: %s \n", outfile);
	/*
	 * sc
	 */
/*	if (write_list(outfile_p, best, distrib) != 0) { */
	if (write_list(outfile_p, best, distrib, section_list, args, argsCount) != 0) {
	fprintf(stderr, "Error writing output file. Exiting.");
		exit(1);
	}
	/* close the file */
	fclose(outfile_p);

	/* We write out the top servers to a file if asked. Note there's no
	   point in freeing the "best" structures. */
	if (topfile) {
	        printf("writing topfile: %s\n", topfile);

		if (write_top(infile_p, topfile_p, best) != 0) {
			fprintf(stderr,
			"Error writing top servers list. Exiting.");
			exit(1);
		}
		/* close the file */
		fclose(topfile_p);
	}
	/* We're all done */
	/* free(best); */
	/* See above: 'there's no point in freeing the "best" structures' */
	/* Bis repetita placent */

	exit(0);
}

void usage()
{
	printf("Usage: apt-spy [options]\n"
		"options:\n"
		"-d distribution\tDebian distribution (ie, stable). Required unless updating.\n"
		"-a area\t\tArea to benchmark. (eg, Europe).\n"
		"-c config\tConfiguration file to use.\n"
		"-e number\tNumber of servers to benchmark before exiting.\n"
		"-y section-list\tComma separated list of section other then main to include.\n");
	printf("-f file\t\tFile to grab when benchmarking. (relative to Debian base).\n"
		"-i file\t\tSpecify input file. For use with the -w option.\n"
		"-m mirror-list\tMirror list to use, or mirror-list to update when updating.\n"
		"-o output-file\tWhere to put output.\n"
		"-p proxy\tProxy server to use. In format <server>:[port]\n"
		"-s country_list\tList of countries to benchmark. Cannot be used with -a.\n"
		"-t time\t\tTime to benchmark each server for. An approximate value only.\n"
		"-u update-URL\tURL to grab mirror-list from when updating.\n");
	printf("-w file\t\tOutput top servers (5 by default) to file for use with -i.\n"
		"-n number\tSpecifies number of top servers to output with -w.\n"
		"-v\t\tOutput a version number.\n"
		"-h\t\tDisplay this message.\n"
		"update\t\tUpdate the mirror list.\n");
	exit(0);
}

void version()
{
	printf("apt-spy %s\n", apt_spy_v);
	exit(0);
}


update.c
/*
 * apt-spy (c) Steven Holmes, 2003.
 * This software is licensed as detailed in the COPYRIGHT file
 */

#include <stdio.h>
#include <curl/curl.h>

#include "include/update.h"

int update(FILE *mirror_p, char *update_url, char *proxy)
{
	CURL *curl;
	char errorbuff[CURL_ERROR_SIZE];
	int error;

	curl = curl_easy_init();

	if (curl == NULL)
		return 1;

	if (proxy != NULL)
		if (curl_easy_setopt(curl, CURLOPT_PROXY, proxy) != 0)
			return 1;

	/* Turn off libcurl's progress indicator */
	if (curl_easy_setopt(curl, CURLOPT_NOPROGRESS, 1) != 0)
		return 1;

	/* Fail on error */
	if (curl_easy_setopt(curl, CURLOPT_FAILONERROR, 1) != 0)
		return 1;

	/* And give us nice, human-readable error messages */
	if (curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, errorbuff) != 0)
		return 1;

	/* set URL to use */
	if (curl_easy_setopt(curl, CURLOPT_URL, update_url) != 0)
		return 1;

	/* set file to output stuff to */
	if (curl_easy_setopt(curl, CURLOPT_FILE, mirror_p) != 0)
		return 1;

	printf("Updating...\n");
	printf("Grabbing file %s...\n", update_url);

	/* Do the transfer */
	if ((error = curl_easy_perform(curl)) != 0) {
		fprintf(stderr, "Error: %i: %s\n", error, errorbuff);
		return 1;
	}

	/* And cleanup */
	curl_easy_cleanup(curl);

	return 0;
}


protocol.c
/*
 * apt-spy (c) Steven Holmes, 2003.
 * This software is licensed as detailed in the COPYRIGHT file
 */

#include <string.h>
#include <assert.h>
#include <stdlib.h>
#include <curl/curl.h>

#include "include/parse.h"
#include "include/protocols.h"

int get_file(server_t *current,
             CURL *curl, char *file,
             enum protocol protocol,
             int *total_bytes)
{
	char errorbuff[CURL_ERROR_SIZE + 1];
	int urlsize=0;
	char *url;
	double download_speed;
	double download_time;
	int error;

	/* Reset byte counter */
	*total_bytes = 0;

	/* Give us human readable error messages in errorbuff */
	if (curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, errorbuff) != 0)
		return 1;

	assert(HTTP==protocol || FTP==protocol);
	urlsize=8+strlen(current->hostname)+strlen(current->path[protocol])+strlen(file);
	url=malloc(urlsize);
	if (!url) {
		perror("malloc");
		exit(1);
	}

	if (HTTP==protocol)	strcpy(url, "http://");
	else if (FTP==protocol)	strcpy(url, "ftp://");

	strcat(url, current->hostname);

	/* Complete the URL. Also save it for later use when writing sources.list */
	if (protocol == HTTP) {
		strcat(url, current->path[HTTP]);
		current->url[HTTP]=strdup(url);
	} else {
		strcat(url, current->path[FTP]);
		current->url[FTP]=strdup(url);
	}

	strcat(url, file);

	/* And register it */
	if (curl_easy_setopt(curl, CURLOPT_URL, url) != 0) {
		free(url);
		return 1;
	}

	/* Now do the actual transfer */
	error = curl_easy_perform(curl);
	if (error != 0) {
		if (error != CURLE_OPERATION_TIMEDOUT) {
			free(url);
			fprintf(stderr, "\t\tError: %s\n", errorbuff);
			return 1;
		} else if (*total_bytes == 0) {
			free(url);
			fprintf(stderr, "\t\tError: %s\n", errorbuff);
			return 1;
		}
	}

	/* And get the statistics and store them */
	curl_easy_getinfo(curl, CURLINFO_TOTAL_TIME, &download_time);

	if (*total_bytes == 0) {
		fprintf(stderr, "\t\tNo data was transfered\n");
		free(url);
		return 1;
	}

	printf("\t\tDownloaded %i bytes in %2.2f seconds\n", *total_bytes, download_time);

	/* Calculate download speed... */
	download_speed = (double) (*total_bytes) / download_time;

	printf("\t\tDownload speed: %.2f kB/sec\n", download_speed / (double) 1024);

	if (download_speed > current->stats.speed) {
		current->stats.speed = download_speed;
		current->stats.protocol = protocol;
	}

	free(url);
	return 0;
}


benchmark.c
/*
 * apt-spy (c) Steven Holmes, 2003.
 *
 * This software is licensed as detailed in the COPYRIGHT file
 */

#include <stdio.h>
#include <string.h>
#include <curl/curl.h>

#include "include/parse.h"
#include "include/benchmark.h"
#include "include/protocols.h"
#include "include/global.h"

/*
 * It is safer to keep track of the total amount of data read ourselves.
 * This fixes a bug appearing when libcurl would not reset its own byte counter
 */
/* static size_t total_bytes = 0; */
static int total_bytes = 0;

/**
 *
 *
 * @param current
 * @param proxy
 * @param timeout
 * @param file
 *
 * @return
 */
int benchmark(server_t *current, char *proxy,  int timeout, char *file)
{

	CURL *curl;

	printf("\nSERVER:\t%s\n", current->hostname);

	curl_global_init (CURL_GLOBAL_ALL);
	/* We use libcurl - here we setup some global options */
	curl = curl_easy_init();

	if (curl == NULL)
		return 1;

	/* Set proxy server to use */
	if (proxy != NULL)
		if (curl_easy_setopt(curl, CURLOPT_PROXY, proxy) != 0)
			return 1;

	/* Connection timeout */
	if (curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, (long) timeout) != 0)
		return 1;

	/* Total timeout */
	if (curl_easy_setopt(curl, CURLOPT_TIMEOUT, (long) timeout) != 0)
		return 1;

	/* Select callback function to deal with data. */
	if (curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, null_write_function) != 0)
		return 1;

	/* Turn off libcurl's progress indicator */
	if (curl_easy_setopt(curl, CURLOPT_NOPROGRESS, 1) != 0)
		return 1;

	/* Fail on error */
	if (curl_easy_setopt(curl, CURLOPT_FAILONERROR, 1) != 0)
		return 1;


	/* Test for an FTP entry */
	if (strlen(current->path[FTP]) != 0) {
		printf("Benchmarking FTP...\n");
		get_file(current, curl, file, FTP, &total_bytes);
	}

	/* Test for an HTTP entry */
	else if (strlen(current->path[HTTP]) != 0) {
		printf("Benchmarking HTTP...\n");
		get_file(current, curl, file, HTTP, &total_bytes);
	} else {
		printf("UNKNOWN TYPE\n");
	}

	curl_easy_cleanup(curl);

	return 0;
}

/**
 *
 *
 * @param current
 * @param best
 */
void decide_best(server_t *current, server_t *best)
{
	int i, j;

	/* Inefficient sorting algorithm, but small number of entries so
	   it doesn't matter. */

	/* move 'i' to the correct place in the array to place the new entry */
	for (i = 0; i < bestnumber; ++i)
		if (current->stats.speed > best[i].stats.speed)
			break;

	/* shove everything along one */
	for (j = bestnumber - 2; j >= i; --j)
		memcpy(&best[j + 1], &best[j], sizeof(server_t));

	/* copy the new entry into the correct place */
	memcpy(&best[i], current, sizeof(server_t));

}

/**
 * @brief Callback from libcurl to allow us to save data. We discard it and
 *        return success.
 *
 * @param bytes
 * @param size
 * @param nmemb
 * @param stream
 *
 * @return
 */
size_t null_write_function(double *bytes, size_t size, size_t nmemb, void *stream)
{
	/* We update total_bytes */
	total_bytes += size * nmemb;
	return size * nmemb;
}




include/parse.h
/*
 * apt-spy (c) Steven Holmes, 2003.
 * This software is licensed as detailed in the COPYRIGHT file.
 */

#ifndef __PARSE_H
#define __PARSE_H

enum protocol { FTP, HTTP };

/* hack */
/*extern int BESTNUMBER;
 */

struct stats_struct {
	int protocol;
	double speed;
};

struct server_struct {
	char	*hostname;
	char	*path[2];
	char	*url[2];
	struct	stats_struct stats;
};

typedef struct server_struct server_t;

char *str_toupper(char *str);
int build_area_file(FILE *config_p, FILE *infile_p, FILE *mirror_list,
					char *area);
int build_country_file(FILE *config_p, FILE *infile_p, FILE *mirror_p,
					   char *country_list);
int find_country(FILE *mirror_list, char *country_code);
void tokenise(server_t *current, char *cur_entry);
char *get_mirrors(FILE *mirror_list);
int write_list(FILE *outfile_p, server_t *best, char *dist, char *section_list, char *args[], int l);
int write_top(FILE *infile_p, FILE *outfile_p, server_t *best);

#endif


trim.c
/*
 * (c) 2009, Stefano Canepa <sc@linux.it>
 *
 * This software is licensed as detailed in the COPYRIGHT file
 *
 *
 * This file license has an history.
 * It was published by mistake with the following copyright note
 *
 * trim.c --- short description
 *
 * Copyright  (C)  2009  Stefano Canepa <sc@linux.it>
 *
 * Version: 1.0
 * Keywords:
 * Author: Stefano Canepa <sc@linux.it>
 * Maintainer: Stefano Canepa <sc@linux.it>
 * URL: http://
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY	; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Commentary:
 * The following code is a modification of the Util_ltrim, Util_rtrim
 * and Util_trim functions of util.{c,h} of the monit project
 * http://savannah.nongnu.org/projects/monit/ released with the same
 * license as this project
 *
 * Then after Paul Wise pointed out in his email on debian-mentors mailing
 * list:  http://lists.debian.org/debian-mentors/2011/12/msg00433.html
 *
 * I requested copyright holders the permission to use it with the following
 * email:

 From: Stefano Canepa <sc@linux.it>
 To: Jan-Henrik Haukeland <hauk@tildeslash.com>,
 Martin Pala <martinp@tildeslash.com>,
 Christian Hopp <chopp@iei.tu-clausthal.de>,
 Michael Amster <mamster@webeasy.com>
 Subject: Need your help
 Organization: Linux.it + ALID.it
 User-Agent: Gnus/5.13 (Gnus v5.13) Emacs/23.3 (gnu/linux)
 Date: Sun, 25 Dec 2011 23:36:47 +0100
 Message-ID: <87liq0s2gw.fsf@lepre.stefanocanepa.home>
 MIME-Version: 1.0
 Content-Type: text/plain; charset=utf-8
 Content-Transfer-Encoding: quoted-printable

 Dear Sirs,
 I'm writing to you becouse I must admit I used your code in a wrong way
 by mistake, but I did.

 I'm trying to maintain apt-spy that is a little program that does some
 tests on debian mirrors and then writes a source file with the fastest
 mirror in it. The code I contribute to this program is not so much for
 me to retain the copyright so I must comply with old developer decision
 to have the code released under BSD three clause lincese.

 Indeed I did two mistakes:
 1) I used some of your code, with very little modification, from your
 util.h/util.c (ltrim, rtrim, trim) creating trim.{h,c} [1] releasing
 it under GPL but stripping stripped your copyright
 2) I linked this code with a BSD based code using OpenSSL enabled curl.

 Just now, after 2 years, under the guidance of Paul Wise of Debian
 understand this is all wrong.

 The solution to 2) should be to link curl without OpenSSL as apt-spy
 does not need OpenSSL at all

 But solution for 1) is not so easy.

 I understand that you can ask me to remove this portion of code that
 violate your rights and I'll do if you want but I prefer to not write
 this code again and I'm asking you for your permission to use your code
 with correct copyright in my project with correct copyright.

 Please accept my apologies.

 Best regards
 Stefano Canepa

 [1] https://github.com/scanepa/apt-spy/blob/master/trim.c

 *
 * To which I received the following reply:
 *

 Delivered-To: scanepa@gmail.com
 Received: by 10.204.230.132 with SMTP id jm4cs221281bkb;
 Mon, 26 Dec 2011 04:32:25 -0800 (PST)
 Received: by 10.14.3.200 with SMTP id 48mr9409831eeh.94.1324902742348;
 Mon, 26 Dec 2011 04:32:22 -0800 (PST)
 Received-SPF: softfail (google.com: best guess record for domain of transitioning
 hauk@tildeslash.com does not designate 62.109.39.247 as permitted sender)
 client-ip=62.109.39.247;
 Received: by 10.14.153.77 with POP3 id e53mf10227544eek.29;
 Mon, 26 Dec 2011 04:32:22 -0800 (PST)
 X-Gmail-Fetch-Info: sc@linux.it 1 picard.linux.it 995 sc
 Return-Path: <hauk@tildeslash.com>
 X-Original-To: sc@linux.it
 Delivered-To: sc@vmbox.linux.it
 Received: from www.tildeslash.com (tildeslash.com [62.109.39.247])
 by picard.linux.it (Postfix) with ESMTP id 0AC263E64AA
 for <sc@linux.it>; Mon, 26 Dec 2011 13:30:13 +0100 (CET)
 Received: by www.tildeslash.com (Postfix, from userid 1003)
 id 420631D02AC; Mon, 26 Dec 2011 13:30:12 +0100 (CET)
 X-Spam-Checker-Version: SpamAssassin 3.2.5 (2008-06-10) on picard.linux.it
 X-Spam-Level: **
 X-Spam-Status: No, score=2.2 required=5.0 tests=DEAR_SOMETHING
 autolearn=disabled version=3.2.5
 Received: from [192.168.10.101] (120.79-161-188.customer.lyse.net [79.161.188.120])
 by www.tildeslash.com (Postfix) with ESMTPSA id 95EDA1D026A;
 Mon, 26 Dec 2011 13:30:02 +0100 (CET)
 Subject: Re: Need your help
 Mime-Version: 1.0 (Apple Message framework v1251.1)
 Content-Type: text/plain; charset=iso-8859-1
 From: Jan-Henrik Haukeland <hauk@tildeslash.com>
 In-Reply-To: <87liq0s2gw.fsf@lepre.stefanocanepa.home>
 Date: Mon, 26 Dec 2011 13:30:02 +0100
 Cc: Martin Pala <martinp@tildeslash.com>,
 Christian Hopp <chopp@iei.tu-clausthal.de>
 Content-Transfer-Encoding: quoted-printable
 Message-Id: <3483D682-2F4C-414F-9EC4-4AE2A8DF26EA@tildeslash.com>
 References: <87liq0s2gw.fsf@lepre.stefanocanepa.home>
 To: Stefano Canepa <sc@linux.it>
 X-Mailer: Apple Mail (2.1251.1)

 Permission granted. You may also change the license to BSD for this =
 particular code if you prefer.=20

 Best regards
 --
 Jan-Henrik Haukeland
 http://tildeslash.com/=20

*/

/**
 * Remove leading char from string
 *
 * @param s A string
 * @param d The char to remove
 *
 * @return s with leading spaces removed
 */

char *ltrim(char *s, const int d)
{
	 char *t= s;

	 if (d == ' ')
		  while(*t==' ' || *t=='\t' || *t=='\r' || *t=='\n') t++;
	 else
		  while(*t==d) t++;
	 if(t!=s) {
		  char *r= s;
		  do {
			   *r++= *t;
		  }
		  while(*t++);
	 }
	 return s;
}

/**
 * Remove trailing char from the string
 *
 * @param s A string
 * @param d The char to remove
 *
 * @return s with trailing spaces removed
 */
char *rtrim(char *s, const int d) {

	char *t= s;


	while(*s) s++;
	if (d == ' ')
		while(*--s==' ' || *s=='\t' || *s=='\r' || *s=='\n') *s= 0;
	else
		while(*--s==d) *s= 0;
	return t;

}

/**
 * Remove leading and trailing space from the string
 * @param s A string
 * @return s with leading and trailing spaces removed
 */
char *trim(char *s, const char d) {

	char *t,*r;

	t = ltrim(s,d);
	r= rtrim(t,d);

	return r;

}
